### ISSUE TITLE
OTA update error stream timeout

### ISSUE THREAD

**Issue by zark-zeugan** [espressif/arduino-esp32#6254]
### Board

ESP32 Dev Model

### Device Description

ESP32 with a attached SIM800L module.

### Hardware Configuration

CPU Freq: 240MHZ (WiFi/BT)
Flash Freq: 80MHz
Flash size: 4MB (32Mb)
Partition scheme: Default 4MB with spiffs
PSRAM: Disabled

### Version

v2.0.2

### IDE Name

Arduino IDE

### Operating System

Windows 10

### Flash frequency

80MHz

### PSRAM enabled

no

### Upload speed

921600

### Description

I am trying to use the on board GSM module to request a binary file from the firebase cloud storage to perfom an OTA update to my device. The size of the binary file is 599904bytes. I am able to receive this data. However, when I try to update using the command `size_t written = Update.writeStream(client);` I get the error `Error 6: UPDATE_ERROR_STREAM` which is a stream read timeout error. 
I am not sure what exactly to make out of this or how this could be solved. 

### Sketch

```cpp
#include <Wire.h>
#include <SPIFFS.h>
#include <ArduinoJson.h>
#include <Firebase_ESP_Client.h>
#include <ArduinoOTA.h>
#include <HTTPUpdate.h>
#include <Update.h>   

// ESP32 LilyGo-T-Call-SIM800 SIM800L_IP5306_VERSION_20190610 (v1.3) pins definition

#define MODEM_UART_BAUD 115200
#define MODEM_RST 5
#define MODEM_PWRKEY 4
#define MODEM_POWER_ON 23
#define MODEM_TX 27
#define MODEM_RX 26
#define I2C_SDA 21
#define I2C_SCL 22
#define LED_PIN 13
#define IP5306_ADDR 0x75
#define IP5306_REG_SYS_CTL0 0x00
#define TOUCHPIN 2


// Set serial for debug console (to the Serial Monitor)
#define SerialMon Serial
// Set serial for AT commands (to the SIM800 module)
#define SerialAT Serial1

// See all AT commands, if wanted
#define DUMP_AT_COMMANDS

// Define the serial console for debug prints, if needed
#define TINY_GSM_DEBUG SerialMon


// See all AT commands, if wanted
#define DUMP_AT_COMMANDS

#define TINY_GSM_USE_GPRS true
#define TINY_GSM_USE_WIFI false

// Configure TinyGSM library
#define TINY_GSM_MODEM_SIM800   // Modem is SIM800
#define TINY_GSM_RX_BUFFER 4026 // Set RX buffer to 1Kb

// Include after TinyGSM definitions
#include <TinyGsmClient.h>

// Your GPRS credentials (leave empty, if missing)
const char apn[] = "airtelgprs.com";    // Your APN
const char gprs_user[] = ""; // User
const char gprs_pass[] = ""; // Password
const char simPIN[] = "";    // SIM card PIN code, if any
#define GSM_PIN ""

// Server details
const char server[]   = "www.google.com";
const char hostname[] = "engeletron-cloud-default-rtdb.firebaseio.com";
const char resource[] = "/devices.json";
const char fota_teste[] = "/fota/teste.json";    // This line added -----------
char data[1024];
int port = 443;
const char* ch;
bool resetflag = false;
long contentLength = 0;
bool isValidContentType = false;

FirebaseConfig config;

const char* FirmwareVer="1.0";
const char* update_resource;
String firmwareURl;
String path_version = "/fota/teste/version";
String path_firmware = "/fota/teste/firmware";

int cnt;
String value_string;

// Layers stack

TinyGsm sim_modem(SerialAT);
TinyGsmClientSecure client(sim_modem);
DynamicJsonDocument doc(1024);


// Modem initial setup (cold start)
void setupModem()
{
  pinMode(MODEM_RST, OUTPUT);
  pinMode(MODEM_PWRKEY, OUTPUT);
  pinMode(MODEM_POWER_ON, OUTPUT);
  pinMode(LED_PIN, OUTPUT);
  pinMode(TOUCHPIN,INPUT_PULLUP);
  
  // Reset pin high
  digitalWrite(MODEM_RST, HIGH);

  // Turn on the Modem power first
  digitalWrite(MODEM_POWER_ON, HIGH);

  // Pull down PWRKEY for more than 1 second according to manual requirements
  digitalWrite(MODEM_PWRKEY, HIGH);
  delay(200);
  digitalWrite(MODEM_PWRKEY, LOW);
  delay(1200);
  digitalWrite(MODEM_PWRKEY, HIGH);

  // Initialize the indicator as an output
  digitalWrite(LED_PIN, LOW);
}


void checkFirmware()
{
  DynamicJsonDocument elements(1024);
  bool startPrint = false;
  int i = 0;

  client.connect(hostname,port,30);
  if(client.connected())
  {
    Serial.println("*************************");
    Serial.println("\n Checking for new Updates ...");
    client.print(String("GET ") + fota_teste + " HTTP/1.1\r\n");
    client.print(String("Host: ") + hostname + "\r\n");
    client.print("Accept: application/json \r\n");
    client.println();
    unsigned long timeout = millis();
    while (client.available() == 0) {
      if (millis() - timeout > 5000) {
        Serial.println("Client Timeout !");
        client.stop();
        return;
      }
    }
    Serial.println("\n Data Received - ");
    while (client.available()) {
      char c = client.read();
      if(c == '{' || startPrint)
      {
        startPrint = true;
        data[i] = c;
        Serial.print(c);
        i++;
      }
    }
  }
  else
  {
    Serial.println("Not connected to the host!!");
  }
  deserializeJson(elements,data);
  const char* ch = elements["version"];
  String result(ch);
  Serial.println("\n \n Current version - " + result);
  if(strcmp(ch,FirmwareVer))
  {
    Serial.println("--- Firmware version not matching ---");
    update_resource = elements["firmware"];
    client.stop();

    exec_Update();
    // execUpdate();
  }

}

String getHeaderValue(String header, String headerName) {
  return header.substring(strlen(headerName.c_str()));
}

void exec_Update()
{
  Serial.println("Downloading binaries ...");
  const char up_hostname[] = "firebasestorage.googleapis.com";
  client.connect(up_hostname,port,30);
  if(!client.connected())
  {
    Serial.println(" >>> Failed <<<");
    return;
  }
    client.print(String("GET ") + String(update_resource) + " HTTP/1.1\r\n");
    client.print(String("Host: ") + up_hostname + "\r\n");
    client.print("Connection: close\r\n");
    client.println();
    //client.println();
    unsigned long timeout = millis();
    while (client.available() == 0) {
      if (millis() - timeout > 5000) {
        Serial.println("Client Timeout !");
        client.stop();
        return;
      }
    }
    while(client.available())
    {
      String line = client.readStringUntil('\n'); //read till '\n'
      line.trim();  // remove space, to check if the line is end of header.
      if (!line.length()) 
      {
        // if line is empty, tis is end of headers
        break;  // Get OTA started after this..
      }   
      if(line.startsWith("HTTP/1.1"))
      {
        if(line.indexOf("200") < 0)
        {
          Serial.println(">>> Got a non 200 status code from the server. Exiting OTA Update. <<<");
          break;
        }
      }
      // Take the content length.
      if(line.startsWith("Content-Length: "))
      {
        contentLength = atol((getHeaderValue(line, "Content-Length: ")).c_str());
        Serial.println("Got " + String(contentLength) + "bytes from server.");
      }
      // Check content type.
      if(line.startsWith("Content-Type: "))
      {
        String contentType = getHeaderValue(line, "Content-Type: ");
        Serial.println("Got " + contentType + " payload.");
        if (contentType == "application/octet-stream")
        {
          isValidContentType = true;
        }
      }
   
    }
    
    Serial.println("Content Length: " + String(contentLength) + ", isValidContentType: " + String(isValidContentType));
    if (contentLength && isValidContentType)
    {
      // Check if enough to perform OTA update.
      bool canBegin = Update.begin(contentLength);
      if(canBegin)
      {
        Serial.println("\n");
        Serial.println("############################################");
        Serial.println("Update may take 2-3 mins to complete...");
        Serial.println("############################################");
        size_t written = Update.writeStream(client);
        if(written == contentLength)
        {
          Serial.println(">Written : " + String(written) + " successfully!!");          
        }
        else
        {
          Serial.println(">Written only :" + String(written) + "/" + String(contentLength) + ". Retry?");
        }

        if(Update.end())
        {
        Serial.println("\n");
        Serial.println("############################################");
        Serial.println("Update completed!");
        Serial.println("############################################");
          if(Update.isFinished())
          {
            Serial.println("Rebooting ...");
            ESP.restart();
          }
          else
          {
            Serial.println("Update not finished ... Something went wrong!");
          }
        }
        else
        {
          Serial.println("Error Occured. Error #: " + String(Update.getError()));
        }
      }
        else
        {
          Serial.println(">>> Not enough space to begin OTA <<<");
          client.flush();
        }
      }
      else
      { 
        Serial.println("There was no content in the response from the server.");
        client.flush();
      }
}


 
void setup()
{
  SerialMon.begin(115200);
  delay(1000);
  // Set SIM module baud rate and UART pins
  SerialAT.begin(115200, SERIAL_8N1, MODEM_RX, MODEM_TX);

  // SIM modem initial setup
  setupModem();
  
  SerialMon.print("Initializing modem...");
  if (!sim_modem.init())
  {
    SerialMon.print(" Failed... restarting modem...");
    setupModem();
    // Restart takes quite some time
    // Use modem.init() if you don't need the complete restart
    if (!sim_modem.restart())
    {
      SerialMon.println(" Fail... even after restart");
      return;
    }
  }
  SerialMon.println(" OK");

  // General information
  String name = sim_modem.getModemName();
  Serial.println("Modem Name: " + name);
  String modem_info = sim_modem.getModemInfo();
  Serial.println("Modem Info: " + modem_info);

  // Unlock your SIM card with a PIN if needed
  if (strlen(simPIN) && sim_modem.getSimStatus() != 3)
  {
    sim_modem.simUnlock(simPIN);
  }

  // Wait for network availability
  SerialMon.print("Waiting for network...");
  if (!sim_modem.waitForNetwork(240000L))
  {
    SerialMon.println(" fail");
    delay(10000);
    return;
  }
  SerialMon.println(" OK");
  
  // Connect to the GPRS network
  SerialMon.print("Connecting to network...");
  if (!sim_modem.isNetworkConnected())
  {
    SerialMon.println(" fail");
    delay(10000);
    return;
  }
  SerialMon.println(" OK");

  // Connect to APN
  SerialMon.print(F("Connecting to APN: "));
  SerialMon.print(apn);
  if (!sim_modem.gprsConnect(apn, gprs_user, gprs_pass))
  {
    SerialMon.println(" fail");
    delay(10000);
    return;
  }  
  SerialMon.println(" OK");

  // More info..
  Serial.println("");
  String ccid = sim_modem.getSimCCID();
  Serial.println("CCID: " + ccid);
  String imei = sim_modem.getIMEI();
  Serial.println("IMEI: " + imei);
  String cop = sim_modem.getOperator();
  Serial.println("Operator: " + cop);
  IPAddress local = sim_modem.localIP();
  Serial.println("Local IP: " + String(local));
  int csq = sim_modem.getSignalQuality();
  Serial.println("Signal quality: " + String(csq));
  checkFirmware();
}
void loop(){
}
```

### Debug Message

```plain
Initializing modem...[3276] ### TinyGSM Version: 0.11.5
[3277] ### TinyGSM Compiled Module:  TinyGsmClientSIM800
[5390] ### Modem: SIMCOM SIM800L
[5390] ### Modem: SIMCOM SIM800L
 OK
[7419] ### Modem: SIMCOM SIM800L
Modem Name: SIMCOM SIM800L
Modem Info: SIM800 R14.18
Waiting for network... OK
Connecting to network... OK
Connecting to APN: airtelgprs.com[14334] ### Network time and time zone updated.
[14335] ### Daylight savings time state updated.
[19877] ### Network time zone updated.
[19881] ### Network time and time zone updated.
[19882] ### Daylight savings time state updated.
 OK


Operator: AirTel
Signal quality: 22
*************************

 Checking for new Updates ...

 Data Received - 
{"firmware":"/v0/b/engeletron-cloud.appspot.com/o/Teste%2Ffirmware.bin?alt=media&token=4a85d162-dea8-445a-82fe-81e99d9ee15f","version":"1.1"}
 
 Current version - 1.1
--- Firmware version not matching ---
Downloading binaries ...
Got application/octet-stream payload.
Got 599904bytes from server.
Content Length: 599904, isValidContentType: 1


############################################
Update may take 2-3 mins to complete...
############################################
>Written only :12694/599904. Retry?
Error Occured. Error #: 6
```

### I have checked existing issues, online documentation and the Troubleshooting Guide

- [X] I confirm I have checked existing issues, online documentation and Troubleshooting guide.

**Comment by VojtechBartoska** [comment]
Hello, can you please retest this on v2.0.3-rc1?

**Comment by zark-zeugan** [comment]
> Hello, can you please retest this on v2.0.3-rc1?

It worked with an http server instead of an https one. 
What do you mean by v2.0.3-rc1?

**Comment by VojtechBartoska** [comment]
ok, so you solved your issue?

2.0.3-rc1 is the latest release version of Arduino core ESP32: https://github.com/espressif/arduino-esp32/releases/tag/2.0.3-RC1

**Comment by podaen** [comment]
I think it is related to my issue #6664, streams just breaks with the current version of WiFi

**Comment by zark-zeugan** [comment]
> ok, so you solved your issue?
> 
> 2.0.3-rc1 is the latest release version of Arduino core ESP32: https://github.com/espressif/arduino-esp32/releases/tag/2.0.3-RC1

I did try with that but faced the same issues. However, I have resolved the issue now, had to use HTTP instead of HTTPS server. The system is still broken (the stream breaks) for downloading files from  HTTPS servers.

**Comment by podaen** [comment]
In the mean time I have updated my sdk... And I'm sure now I doesn't run in to timeout, but it still breaks for me too. So no solution right now.

**Comment by mudgalp** [comment]
> ### Board
> ESP32 Dev Model
> 
> ### Device Description
> ESP32 with a attached SIM800L module.
> 
> ### Hardware Configuration
> CPU Freq: 240MHZ (WiFi/BT) Flash Freq: 80MHz Flash size: 4MB (32Mb) Partition scheme: Default 4MB with spiffs PSRAM: Disabled
> 
> ### Version
> v2.0.2
> 
> ### IDE Name
> Arduino IDE
> 
> ### Operating System
> Windows 10
> 
> ### Flash frequency
> 80MHz
> 
> ### PSRAM enabled
> no
> 
> ### Upload speed
> 921600
> 
> ### Description
> I am trying to use the on board GSM module to request a binary file from the firebase cloud storage to perfom an OTA update to my device. The size of the binary file is 599904bytes. I am able to receive this data. However, when I try to update using the command `size_t written = Update.writeStream(client);` I get the error `Error 6: UPDATE_ERROR_STREAM` which is a stream read timeout error. I am not sure what exactly to make out of this or how this could be solved.
> 
> ### Sketch
> ```c++
> #include <Wire.h>
> #include <SPIFFS.h>
> #include <ArduinoJson.h>
> #include <Firebase_ESP_Client.h>
> #include <ArduinoOTA.h>
> #include <HTTPUpdate.h>
> #include <Update.h>   
> 
> // ESP32 LilyGo-T-Call-SIM800 SIM800L_IP5306_VERSION_20190610 (v1.3) pins definition
> 
> #define MODEM_UART_BAUD 115200
> #define MODEM_RST 5
> #define MODEM_PWRKEY 4
> #define MODEM_POWER_ON 23
> #define MODEM_TX 27
> #define MODEM_RX 26
> #define I2C_SDA 21
> #define I2C_SCL 22
> #define LED_PIN 13
> #define IP5306_ADDR 0x75
> #define IP5306_REG_SYS_CTL0 0x00
> #define TOUCHPIN 2
> 
> 
> // Set serial for debug console (to the Serial Monitor)
> #define SerialMon Serial
> // Set serial for AT commands (to the SIM800 module)
> #define SerialAT Serial1
> 
> // See all AT commands, if wanted
> #define DUMP_AT_COMMANDS
> 
> // Define the serial console for debug prints, if needed
> #define TINY_GSM_DEBUG SerialMon
> 
> 
> // See all AT commands, if wanted
> #define DUMP_AT_COMMANDS
> 
> #define TINY_GSM_USE_GPRS true
> #define TINY_GSM_USE_WIFI false
> 
> // Configure TinyGSM library
> #define TINY_GSM_MODEM_SIM800   // Modem is SIM800
> #define TINY_GSM_RX_BUFFER 4026 // Set RX buffer to 1Kb
> 
> // Include after TinyGSM definitions
> #include <TinyGsmClient.h>
> 
> // Your GPRS credentials (leave empty, if missing)
> const char apn[] = "airtelgprs.com";    // Your APN
> const char gprs_user[] = ""; // User
> const char gprs_pass[] = ""; // Password
> const char simPIN[] = "";    // SIM card PIN code, if any
> #define GSM_PIN ""
> 
> // Server details
> const char server[]   = "www.google.com";
> const char hostname[] = "engeletron-cloud-default-rtdb.firebaseio.com";
> const char resource[] = "/devices.json";
> const char fota_teste[] = "/fota/teste.json";    // This line added -----------
> char data[1024];
> int port = 443;
> const char* ch;
> bool resetflag = false;
> long contentLength = 0;
> bool isValidContentType = false;
> 
> FirebaseConfig config;
> 
> const char* FirmwareVer="1.0";
> const char* update_resource;
> String firmwareURl;
> String path_version = "/fota/teste/version";
> String path_firmware = "/fota/teste/firmware";
> 
> int cnt;
> String value_string;
> 
> // Layers stack
> 
> TinyGsm sim_modem(SerialAT);
> TinyGsmClientSecure client(sim_modem);
> DynamicJsonDocument doc(1024);
> 
> 
> // Modem initial setup (cold start)
> void setupModem()
> {
>   pinMode(MODEM_RST, OUTPUT);
>   pinMode(MODEM_PWRKEY, OUTPUT);
>   pinMode(MODEM_POWER_ON, OUTPUT);
>   pinMode(LED_PIN, OUTPUT);
>   pinMode(TOUCHPIN,INPUT_PULLUP);
>   
>   // Reset pin high
>   digitalWrite(MODEM_RST, HIGH);
> 
>   // Turn on the Modem power first
>   digitalWrite(MODEM_POWER_ON, HIGH);
> 
>   // Pull down PWRKEY for more than 1 second according to manual requirements
>   digitalWrite(MODEM_PWRKEY, HIGH);
>   delay(200);
>   digitalWrite(MODEM_PWRKEY, LOW);
>   delay(1200);
>   digitalWrite(MODEM_PWRKEY, HIGH);
> 
>   // Initialize the indicator as an output
>   digitalWrite(LED_PIN, LOW);
> }
> 
> 
> void checkFirmware()
> {
>   DynamicJsonDocument elements(1024);
>   bool startPrint = false;
>   int i = 0;
> 
>   client.connect(hostname,port,30);
>   if(client.connected())
>   {
>     Serial.println("*************************");
>     Serial.println("\n Checking for new Updates ...");
>     client.print(String("GET ") + fota_teste + " HTTP/1.1\r\n");
>     client.print(String("Host: ") + hostname + "\r\n");
>     client.print("Accept: application/json \r\n");
>     client.println();
>     unsigned long timeout = millis();
>     while (client.available() == 0) {
>       if (millis() - timeout > 5000) {
>         Serial.println("Client Timeout !");
>         client.stop();
>         return;
>       }
>     }
>     Serial.println("\n Data Received - ");
>     while (client.available()) {
>       char c = client.read();
>       if(c == '{' || startPrint)
>       {
>         startPrint = true;
>         data[i] = c;
>         Serial.print(c);
>         i++;
>       }
>     }
>   }
>   else
>   {
>     Serial.println("Not connected to the host!!");
>   }
>   deserializeJson(elements,data);
>   const char* ch = elements["version"];
>   String result(ch);
>   Serial.println("\n \n Current version - " + result);
>   if(strcmp(ch,FirmwareVer))
>   {
>     Serial.println("--- Firmware version not matching ---");
>     update_resource = elements["firmware"];
>     client.stop();
> 
>     exec_Update();
>     // execUpdate();
>   }
> 
> }
> 
> String getHeaderValue(String header, String headerName) {
>   return header.substring(strlen(headerName.c_str()));
> }
> 
> void exec_Update()
> {
>   Serial.println("Downloading binaries ...");
>   const char up_hostname[] = "firebasestorage.googleapis.com";
>   client.connect(up_hostname,port,30);
>   if(!client.connected())
>   {
>     Serial.println(" >>> Failed <<<");
>     return;
>   }
>     client.print(String("GET ") + String(update_resource) + " HTTP/1.1\r\n");
>     client.print(String("Host: ") + up_hostname + "\r\n");
>     client.print("Connection: close\r\n");
>     client.println();
>     //client.println();
>     unsigned long timeout = millis();
>     while (client.available() == 0) {
>       if (millis() - timeout > 5000) {
>         Serial.println("Client Timeout !");
>         client.stop();
>         return;
>       }
>     }
>     while(client.available())
>     {
>       String line = client.readStringUntil('\n'); //read till '\n'
>       line.trim();  // remove space, to check if the line is end of header.
>       if (!line.length()) 
>       {
>         // if line is empty, tis is end of headers
>         break;  // Get OTA started after this..
>       }   
>       if(line.startsWith("HTTP/1.1"))
>       {
>         if(line.indexOf("200") < 0)
>         {
>           Serial.println(">>> Got a non 200 status code from the server. Exiting OTA Update. <<<");
>           break;
>         }
>       }
>       // Take the content length.
>       if(line.startsWith("Content-Length: "))
>       {
>         contentLength = atol((getHeaderValue(line, "Content-Length: ")).c_str());
>         Serial.println("Got " + String(contentLength) + "bytes from server.");
>       }
>       // Check content type.
>       if(line.startsWith("Content-Type: "))
>       {
>         String contentType = getHeaderValue(line, "Content-Type: ");
>         Serial.println("Got " + contentType + " payload.");
>         if (contentType == "application/octet-stream")
>         {
>           isValidContentType = true;
>         }
>       }
>    
>     }
>     
>     Serial.println("Content Length: " + String(contentLength) + ", isValidContentType: " + String(isValidContentType));
>     if (contentLength && isValidContentType)
>     {
>       // Check if enough to perform OTA update.
>       bool canBegin = Update.begin(contentLength);
>       if(canBegin)
>       {
>         Serial.println("\n");
>         Serial.println("############################################");
>         Serial.println("Update may take 2-3 mins to complete...");
>         Serial.println("############################################");
>         size_t written = Update.writeStream(client);
>         if(written == contentLength)
>         {
>           Serial.println(">Written : " + String(written) + " successfully!!");          
>         }
>         else
>         {
>           Serial.println(">Written only :" + String(written) + "/" + String(contentLength) + ". Retry?");
>         }
> 
>         if(Update.end())
>         {
>         Serial.println("\n");
>         Serial.println("############################################");
>         Serial.println("Update completed!");
>         Serial.println("############################################");
>           if(Update.isFinished())
>           {
>             Serial.println("Rebooting ...");
>             ESP.restart();
>           }
>           else
>           {
>             Serial.println("Update not finished ... Something went wrong!");
>           }
>         }
>         else
>         {
>           Serial.println("Error Occured. Error #: " + String(Update.getError()));
>         }
>       }
>         else
>         {
>           Serial.println(">>> Not enough space to begin OTA <<<");
>           client.flush();
>         }
>       }
>       else
>       { 
>         Serial.println("There was no content in the response from the server.");
>         client.flush();
>       }
> }
> 
> 
>  
> void setup()
> {
>   SerialMon.begin(115200);
>   delay(1000);
>   // Set SIM module baud rate and UART pins
>   SerialAT.begin(115200, SERIAL_8N1, MODEM_RX, MODEM_TX);
> 
>   // SIM modem initial setup
>   setupModem();
>   
>   SerialMon.print("Initializing modem...");
>   if (!sim_modem.init())
>   {
>     SerialMon.print(" Failed... restarting modem...");
>     setupModem();
>     // Restart takes quite some time
>     // Use modem.init() if you don't need the complete restart
>     if (!sim_modem.restart())
>     {
>       SerialMon.println(" Fail... even after restart");
>       return;
>     }
>   }
>   SerialMon.println(" OK");
> 
>   // General information
>   String name = sim_modem.getModemName();
>   Serial.println("Modem Name: " + name);
>   String modem_info = sim_modem.getModemInfo();
>   Serial.println("Modem Info: " + modem_info);
> 
>   // Unlock your SIM card with a PIN if needed
>   if (strlen(simPIN) && sim_modem.getSimStatus() != 3)
>   {
>     sim_modem.simUnlock(simPIN);
>   }
> 
>   // Wait for network availability
>   SerialMon.print("Waiting for network...");
>   if (!sim_modem.waitForNetwork(240000L))
>   {
>     SerialMon.println(" fail");
>     delay(10000);
>     return;
>   }
>   SerialMon.println(" OK");
>   
>   // Connect to the GPRS network
>   SerialMon.print("Connecting to network...");
>   if (!sim_modem.isNetworkConnected())
>   {
>     SerialMon.println(" fail");
>     delay(10000);
>     return;
>   }
>   SerialMon.println(" OK");
> 
>   // Connect to APN
>   SerialMon.print(F("Connecting to APN: "));
>   SerialMon.print(apn);
>   if (!sim_modem.gprsConnect(apn, gprs_user, gprs_pass))
>   {
>     SerialMon.println(" fail");
>     delay(10000);
>     return;
>   }  
>   SerialMon.println(" OK");
> 
>   // More info..
>   Serial.println("");
>   String ccid = sim_modem.getSimCCID();
>   Serial.println("CCID: " + ccid);
>   String imei = sim_modem.getIMEI();
>   Serial.println("IMEI: " + imei);
>   String cop = sim_modem.getOperator();
>   Serial.println("Operator: " + cop);
>   IPAddress local = sim_modem.localIP();
>   Serial.println("Local IP: " + String(local));
>   int csq = sim_modem.getSignalQuality();
>   Serial.println("Signal quality: " + String(csq));
>   checkFirmware();
> }
> void loop(){
> }
> ```
> 
> ### Debug Message
> ```
> Initializing modem...[3276] ### TinyGSM Version: 0.11.5
> [3277] ### TinyGSM Compiled Module:  TinyGsmClientSIM800
> [5390] ### Modem: SIMCOM SIM800L
> [5390] ### Modem: SIMCOM SIM800L
>  OK
> [7419] ### Modem: SIMCOM SIM800L
> Modem Name: SIMCOM SIM800L
> Modem Info: SIM800 R14.18
> Waiting for network... OK
> Connecting to network... OK
> Connecting to APN: airtelgprs.com[14334] ### Network time and time zone updated.
> [14335] ### Daylight savings time state updated.
> [19877] ### Network time zone updated.
> [19881] ### Network time and time zone updated.
> [19882] ### Daylight savings time state updated.
>  OK
> 
> 
> Operator: AirTel
> Signal quality: 22
> *************************
> 
>  Checking for new Updates ...
> 
>  Data Received - 
> {"firmware":"/v0/b/engeletron-cloud.appspot.com/o/Teste%2Ffirmware.bin?alt=media&token=4a85d162-dea8-445a-82fe-81e99d9ee15f","version":"1.1"}
>  
>  Current version - 1.1
> --- Firmware version not matching ---
> Downloading binaries ...
> Got application/octet-stream payload.
> Got 599904bytes from server.
> Content Length: 599904, isValidContentType: 1
> 
> 
> ############################################
> Update may take 2-3 mins to complete...
> ############################################
> >Written only :12694/599904. Retry?
> Error Occured. Error #: 6
> ```
> 
> ### Other Steps to Reproduce
> _No response_
> 
> ### I have checked existing issues, online documentation and the Troubleshooting Guide
> * [x]  I confirm I have checked existing issues, online documentation and Troubleshooting guide.

i'm also facing the same issue with quectel's LTE module EC200S.

**Comment by me-no-dev** [comment]
maybe there is a way to set the timeout of the client to something higher? GSM modems would be much slower than WiFi

**Comment by mudgalp** [comment]
@zark-zeugan can you please help me in the same problem, i'm already using http server made by my team and LTE module EC200S from Quectel.

Thank you in advance.

**Comment by VojtechBartoska** [comment]
Hello folks, do you still need help on this?

**Comment by glen020** [comment]
I would like some help. Having the same issue. 
Downloading 768672 Bytes (~745KB)
It takes 959 seconds (~16 mins) to timeout.

**Comment by podaen** [comment]
It seems that you don't empty the buffer engough. I remember doing it in 250ms for 1 mega. Never the less, it is still a hugh package size for a microcontroller to receive. Maybe you have to try with much less, and slitly increase it and see what it does.

**Comment by glen020** [comment]
So I think it might be an Azure issue. Fails in an azure flask webpage BUT works in a "pythonanywhere.com" flask website. Will have to do more digging or hopefully someone here comes across something.

**Comment by VojtechBartoska** [comment]
are there any updates on this guys? Thank you for your answers :)

**Comment by zark-zeugan** [comment]
So, I have tried a lot of things on this and adding buffers or even breaking the data does not help much. The checksum in the end always does come errors. The only resolution I was able to find was to use a HTTP server like the AWS one to host my bin files on and was able to the system working for bigger file sizes of OTA Updates. 
I am not sure as to how much has the firmware of the system been developed to now. But as of now I have concluded that the system can only work with HTTP server and not with HTTPS.
